### ISSUE TITLE
kubernetes support

### ISSUE THREAD

**Issue by pearj** [zalando/zalenium#103]
Hi,

I'm wondering if there are any plans to add kubernetes support?

I realise that zalenium as it stands is very tied to using the docker command directly.  But with a bit of refactoring, I think it wouldn't be too hard to have a zalenium-docker and a zalenium-kubernetes version if the core bits of zalenium were moved to a separate maven artifact and the docker specific parts moved to another maven artefact, which would allow a kubernetes version.

Thoughts?

**Comment by elgalu** [comment]
Hi, as of now we are using https://github.com/spotify/docker-client to interact with Docker.

To work on a kubernetes version I believe we would need a kubernetes-java client, as of now there only seem to exist go and python clients:
https://github.com/kubernetes/client-go
https://github.com/kubernetes-incubator/client-python

**Comment by elgalu** [comment]
Ah, found non-official java clients:

https://github.com/fabric8io/fabric8/tree/master/components/kubernetes-api
https://github.com/fabric8io/kubernetes-client

**Comment by pearj** [comment]
Hi,

I had the fabric8 kubernetes-client in mind as well when I created this issue.
The thing that is cool about it, is that it has support for authenticating via a service account which means that the authentication to the Kubernetes API should be transparent, as long as the zalenium pod is created with a service account that has enough permissions to be able to create and delete pods, and run commands against a pod.

Starting a selenium Pod would probably look something like this:
```java
DefaultKubernetesClient client = new DefaultKubernetesClient();
Pod newPod = client.pods()
    .inNamespace("rio")
    .createNew()
    .withNewMetadata()
        .withGenerateName("selenium-node")
        .addToLabels("app", "zalenium")
    .endMetadata()
    .withNewSpec()
        .addNewContainer()
            .withName("selenium-node")
            .withImage("elgalu/selenium")
            .addNewEnv()
                .withName("Bob")
                .withValue("The builder")
            .endEnv()
        .endContainer()
        .withRestartPolicy("Never")
    .endSpec()
    .done();
```

Does Zalendo use Kubernetes at the moment? Or do you not have any Kubernetes capability at the moment?

**Comment by diemol** [comment]
Zalando is moving to Kubernetes, so we will start checking into this soon. Nevertheless any help is highly appreciated :)

**Comment by dspasojevic** [comment]
@diemol if you are able to provide some rough direction for how you would like docker and kubernetes support to co-exist @pearj and I may be able to action this.

**Comment by diemol** [comment]
Hi @dspasojevic, sorry for the late reply.

Thanks for offering help! We really appreciate that.

The first thing we need to do is abstract the docker calls made through the java docker-client from Spotify. This will be basically an extra class that wraps all the calls. I will work on that during this week.

After that, we need to establish some sort of logic that decides if Zalenium is running only with docker or in k8s. The logic would simply decide which api client to call, either the Spotify docker-client or the fabric8 Kubernetes client. That logic could be in the wrapper class I mentioned above.

My dream would be that we could implement this in a way that we could support Docker Swarm as well, but that is a "nice to have".

@pearj, @dspasojevic how does that sound for you?

**Comment by pearj** [comment]
Hi @diemol, that's sounds good. We should be able to detect kubernetes by for the kubernetes service account which gets mounted in at /var/run/secrets/kubernetes.io

**Comment by diemol** [comment]
Hi @pearj and @dspasojevic,

We just released https://github.com/zalando/zalenium/releases/tag/3.3.1g, and the code contains a refactoring that can be seen in the "container" package. Basically I tried to abstract the docker-client calls.

In theory, we should be able to implement a ContainerClient with Kubernetes, add some logic in the ContainerFactory to instantiate it when we are running k8s. 

I do not discard that maybe we need an extra class that mixes both DockerContainerClient and KubernetesContainerClient to actually create everything in a proper way.

Also for the unit tests I tried to add each ContainerClient as an element for a DataProvider, so we don't need to duplicate the tests but perhaps just reuse the ones we have now.

If you have some time, please check it and let me know your thoughts, hopefully this approach looks ok for you guys.

**Comment by pearj** [comment]
Hi @diemol,
I noticed your code changes, and I spent the day already starting work on it. 
I managed to get the fabric8 kubernetes-client working inside kubernetes (actually Openshift, but it extends kubernetes) using the built-in service account to create a pod (in a standalone test project).  
After that, I started modifying the shell scripts to not crash when starting up inside kubernetes, because the docker client wasn't working.
Tomorrow, I hope to have a crack at implementing the ContainerClient with Kubernetes.  I should at the very least be able to start a pod by the end of tomorrow, should be fun.
Thanks a lot for the work you did last week, I'll let you know how it goes.

**Comment by pearj** [comment]
Hi @diemol and other zalandians,

So, I've hit a bit of a snag. I've realised that zalenium currently uses a special docker network mode called "container" which appears to make all the selenium nodes share the same network interface.  This works fine on the same host, but that technique isn't going to work across separate kubernetes nodes.
I managed to get the selenium nodes registering ok, but I don't have novnc working, as I think it still assumes everything is on localhost.

Do you have any guidance on what is required to make novnc work properly across different IP addresses? Can you give me some background why zalenium was designed this way in the first place? Security?

You can take a look at what I've done so far to get an idea of where I'm heading with it.
https://github.com/IPAustralia/zalenium/tree/kubernetes

Working so far:
* Starting selenium containers from the zalenium pod via the kubernetes-client
* Detecting when running inside kubernetes
* Selenium nodes registering with the zalenium hub
* Deleting stale selenium nodes on startup
* Deleting selenium nodes on JVM shutdown
* Finding the number of running selenium nodes
* Mounting in a Memory volume at /dev/shm 

Not yet working:
* Executing commands on nodes - Looks pretty straight forward from the samples I've seen, just haven't got to it yet
* Copying videos - Looks like executing tar command on the selenium node to stdout and decompressing on the zalenium hub is the easiest way to do this.  That is what kubernetes command line util (kubectl/oc) seems to do when you ask it to copy files around.
* Auto-mounting the mounted folder - This should be relatively straight forward, using the kubernetes client I can lookup the pod details of the zalenium hub and see if there is a mounted folder, then I can use the same persistent volume claim when starting the selenium node
* Novnc not working - need advice as mentioned above
* Pre-downloading the latest selenium image is problematic for kubernetes I believe, in OpenShift you an create an image stream that is based off something from docker hub, and it puts the image in the private registry under the namespace that zalenium is running in.  Not sure if straight kubernetes can do this.  I guess it's not the end of the world, but it just means that it'll be slow to start if you hit a node that hasn't run a selenium node.
* Reading the docker label for the selenium nodes is looking problematic, the kubernetes API doesn't seem to support it, but the OpenShift does have an image API that can find that information, but it's harder than you'd expect to look up details by image name.  You have to list all the images in the private registry and loop through the results. This doesn't help kubernetes though

Issues to think about:
* How often should the selenium node image be refreshed from the docker hub? On every selenium node start might be too often for kubernetes, maybe it needs to happen on a timer? Although this might only be implementable in Openshift...

What flavour of Kubernetes are you guys planning on using? Straight kubernetes or OpenShift Origin? Minikube? Tectonic?

**Comment by diemol** [comment]
Hi @pearj,

First, thanks for the effort! I've taken a look to your branch and it looks great, with all the things you have been working so far! Overall I think we are moving on the right direction.

I realised that we need some more changes, in the startup script and in the concept as well (like the network mode). So, to answer you:

> I've realised that zalenium currently uses a special docker network mode called "container" which appears to make all the selenium nodes share the same network interface. 

We did that for simplicity in that moment, to facilitate the live preview and other small things. But it is time to change it and I will work on that in the next days so we just use containers IPs and the normal bridged network mode. I also hope that this will help #116.

 > What flavour of Kubernetes are you guys planning on using? Straight kubernetes or OpenShift Origin? Minikube? Tectonic?

For local testing Minikube, and in Zalando we are using straight vanilla Kubernetes running on AWS.

To keep moving on, we can do the following:
- I'll work on removing this network container mode, hopefully to have it gone by early next week.
- Then you could merge those changes in your fork and see the current status.
- We find a way to work on the next changes together (perhaps I create a branch here and we work there?), so we figure out a way to make the remaining features work.

What do you think?

Again, thanks for the help and hopefully we can have this working smoothly.

**Comment by pearj** [comment]
> I'll work on removing this network container mode, hopefully to have it gone by early next week.
Then you could merge those changes in your fork and see the current status.

Great sounds good
> We find a way to work on the next changes together (perhaps I create a branch here and we work there?), so we figure out a way to make the remaining features work.

Ok, so you create a kubernetes branch and I'll create a PR to that branch?

> For local testing Minikube, and in Zalando we are using straight vanilla Kubernetes running on AWS.

Ok, so we really need all the core features to use the kubernetes API then. 

I was also thinking after reading the self healing comment about swarm and kubernetes in #116 that maybe we should let kubernetes manage the initial containers in a replication controller. 

That means we can tell kubernetes to make sure we always have 5 selenium pods for example. Then if zalenium decides to kill one of those pods then it will automatically recreate it again. Which could be handy to ensure base capacity that is always available. 
I'm not sure what happens if you manually start up a pod with the same labels as one started by a replication controller, if kubernetes will kill it off if it exceeds the number of replicas. 
If it does it's no issue as we can leave burst capacity outside of a replication controller that zalenium manages on its own.

**Comment by pearj** [comment]
Hi @diemol,

Not sure if it's related to kubernetes, but now that I have novnc working, I've noticed that the aspect ratio is wrong, it's too thin. The videos are recorded that way too. Is this a known issue? Or have I caused this somehow?

**Comment by pearj** [comment]
> Not sure if it's related to kubernetes, but now that I have novnc working, I've noticed that the aspect ratio is wrong, it's too thin. The videos are recorded that way too. Is this a known issue? Or have I caused this somehow?

Turns out if I change the screen resolution it's fine, so it's all good.

**Comment by diemol** [comment]
Exactly, when I saw your comment I thought that right away. Somehow I just forgot to write it down here.
Maybe we should change the default screen resolution? It does look kind of strange. What do you think @elgalu ?

**Comment by elgalu** [comment]
Yes, we could change it something more square-ish so the live preview page looks better.

**Comment by pearj** [comment]
Hi @elgalu @diemol,

I noticed that `de.zalando.ep.zalenium.container.DockerContainerClient.executeCommand(String, String[], boolean)` now has a `waitForExecution`  (the last boolean) option. It look like `de.zalando.ep.zalenium.proxy.DockerSeleniumRemoteProxy.DockerSeleniumContainerAction.START_RECORDING` is the only one that doesn't wait for execution.  Is this to speed up the start of a test executing, by sacrificing the first few seconds of the video?  I hadn't bothered to support this mode yet for kubernetes, as I wasn't sure if I still needed to read all the output and close all the connections for the command to work properly.  So basically `waitForExecution==false` means to be asynchronous I guess?

If you think this is important I'll work out a way to support this in kubernetes.

**Comment by diemol** [comment]
Hey @pearj,

>  Is this to speed up the start of a test executing

Exactly, the idea is to answer back to the grid as fast as possible.

> sacrificing the first few seconds of the video?

I thought the same at the beginning, but after a few tests we saw that no seconds are sacrificed. Starting the video recording doesn't take that long and that time is used to answer back all the way to the client and allow it to send the first command (usually `driver.get()`). 

> So basically `waitForExecution==false` means to be asynchronous I guess?

Yeah.


We can leave it the way it is for now in Kubernetes, and perhaps just leave it as a TODO for the future.

**Comment by pearj** [comment]
I couldn't resist last night and added asynchronous support execute command. It seemed to work ok today. 
I also played with CPU limits today. And I noticed inside the docker container running top that ffmpeg used about a stable 17% of a CPU (we're running our openshift nodes on VMware) but I don't know how that compares with Amazon CPUs. The thing that spiked in CPU usage the most is google chrome.

In a test that I was watching when selenium was waiting for an event it would use about 20-25% of a CPU and when chrome was doing something it would spike to over 100%. 

So I put a CPU request of 25% and a CPU limit of 50% and I confirmed in docker stats that kubernetes was stopping the container going above 50%.

I noticed that zalenium itself didn't spike much above 15%.  So the video recording really does add a fair bit as expected. 

I didn't get access to my development VM today either, so I think I'll give up waiting and start writing the doco again (I didn't get that far) and add the support for overriding the selenium docker image again.

**Comment by diemol** [comment]
That's great @pearj!

I also checked again, and the docker part is still working fine (I made a couple of changes only related to docker).

What if you join the gitter channel and we discuss next steps? I only have in mind the documentation, doing some tests in Kubernetes, and then we could just merge it to master to release in a beta state. What do you think?

**Comment by diemol** [comment]
Closed by #160
